### DISCUSSION TITLE
Show Items ONLY on Reading Mode

### DISCUSSION THREAD

**Discussion by FelipeRearden** [chrisgurney/obsidian-note-toolbar#16] [Ideas]
Hello @chrisgurney !!!!

After using the Note Toolbar on real workflows I realize that would be important to be able to have a different  Note Toolbar when the Active Note is in Reading Mode.

For my Workflow, I would have Items on Reading Mode to open the Right and Left Sidebars with different plugins since I really use Reading Mode to read :)

## My Idea

A new checkbox inside the `Add Item` window under `Hide On` . When the checkbox is marked, it will hide the item when the active is in Reading Mode.

<hr>

Thanks for reading this !

Let me know what you think :)

I wish you a fantastic day ☀️

**Comment by chrisgurney** [comment]
Interesting! 

I find this insightful as I don't use Reading mode at all.

Why do you use Reading mode and can you give some examples of how you would use the toolbar in those cases?

**Comment by FelipeRearden** [comment]
> Why do you use Reading mode and can you give some examples of how you would use the toolbar in those cases?

🚨 This is a WIP that I was working before Note Toolbar. It is  not perfect yet it will help to understand how to use toolbar on RM.

## Core Idea

On reading mode I wanna hide all commands that change the note (bold, italic, highlight) and focus on commands that help me to read and not lost FOCUS and easy to trigger.

## Assumptions

1️⃣ Right and Left Sidebar always open 
2️⃣ One note in the center on Reading Mode
3️⃣ Right Sidebar is a draft note where I write my thoughts while reading
4️⃣ Left Sidebar has plugins that has Sidecar UI: Search, Contextual Sidecar Note* and Outline

Contextual Sidecar Plugin is a plugin that let you visualize a note based on YAML parameter*

https://github.com/matthewturk/obsidian-sidecar-panel

## Toolbar

`Item 1` Search in the file
`Item 2` Open Note Headings
`Item 3` Prompt to Add a content to the Draft Note


## Operation 

🎥 While Reading I can call the Note Toolbar to execute commands 

https://github.com/chrisgurney/obsidian-note-toolbar/assets/85180696/9cd46cdf-e591-4b98-b073-939f7db8aabc

<hr>

The final workflow will be more detailed and with many many more Toolbar item. I create this mock up just to give you the gist of my idea. As I said the main idea is to be able to enhance the Reading experience by adding `tools` to support the reading.

Without Note toolbar I am losing focus remembering the hotkeys. I just need to trigger the Toolbar and execute the command and keep reading 🙏

Let me know if you need more information ☀️

**Comment by chrisgurney** [comment]
This is great, thank you for the insights!

Some first thoughts:

- I think I have to restructure the settings around visibility, which is really just set in two flags right now: hideOnDesktop and hideOnMobile.
  - This might then break into a `visibility` setting that breaks down into a series of flags: `desktop`, `mobile`, `reading`, `editing`.
- I would have to replace the two settings UI toggles with some sort of multi-select UI, to make better use of screen real estate. This could take the form of:
  1. a dropdown with checkboxes (have you seen anything like this in another plugin?)  ...OR
  2. a minimal, textual (horizontal) list of custom UI that looks something like the following (click on each to change) ...OR 
    - Show on: desktop ✓ mobile ✓ reading ✓ editing 𐄂
  3. I could reuse the _Styles_ settings UI (i.e., chips), where all four chips shown by default, OR none are shown, and you select each to hide individually. But the former feels noisy, and the latter is opposite to the way it works in the Styles section  ...OR
  4. some other solution.

**Reply by FelipeRearden** [reply]
> This is great, thank you for the insights!

💙 thank you very much !!!!!

> I would have to replace the two settings UI toggles with some sort of multi-select UI, to make better use of screen real estate. This could take the form of:
>
> i. a dropdown with checkboxes (have you seen anything like this in another plugin?) ...OR

I am trying to think about this and I dont know if I saw any plugin with this type of multilsetlect dropdown. And I dont know if this UI is mobile friendly. 

> a minimal, textual (horizontal) list of custom UI that looks something like the following (click on each to change) ...OR
> Show on: desktop ✓ mobile ✓ reading ✓ editing

Sharing my thoughts: If the problem is UI space you could replace the words for icons. Like this:

![IMG_1161](https://github.com/chrisgurney/obsidian-note-toolbar/assets/85180696/64abaa4e-39ee-4ecf-974b-b7ec6af89714)

Lucide Terms
```
monitor
smartphone
pencil
book-open
```

`pencil` and `book-open` are the native icons used in Obsidian UI to represent `Live Preview` and `Reading View`

**Reply by chrisgurney** [reply]
@FelipeRearden Coming back to this:

> On reading mode I wanna hide all commands that change the note (bold, italic, highlight) and focus on commands that help me to read and not lost FOCUS and easy to trigger.

Do you think you might have different preferences around toolbar items, based on whether you're in Reading mode on desktop versus mobile?

**Reply by FelipeRearden** [reply]
@chrisgurney in my humble opinion, the answer in NO. 

For me on Readign Mode, the only difference between Desktop and Mobile is the size of the screen and the scroll tool (mouse vs tap). 

If the user has an specific case, I use this variable in my templater scripts to manage different behavior for a command between mobile and Desktop in a `if statement`

```
const OS = app.isMobile
```

This way, power user could have a way to differentiate between Mobile and Desktop ☀️

**Comment by chrisgurney** [comment]
Per [this update](https://github.com/chrisgurney/obsidian-note-toolbar/issues/20#issuecomment-2059821079), I've hopefully laid the groundwork for a future "context" setting, which should allow you to set the toolbar position _and_ hide items based on a combination of platform (Desktop / Mobile) and view mode (Reading / Editing).

**Reply by FelipeRearden** [reply]
Hello @chrisgurney !!!!

First of all, forgive me for the late reply. This week has been a hard one but I am gonna use the weekend to test the features from the new version using the version `1.5.15`.

> Per https://github.com/chrisgurney/obsidian-note-toolbar/issues/20#issuecomment-2059821079, I've hopefully laid the groundwork for a future "context" setting, which should allow you to set the toolbar position and hide items based on a combination of platform (Desktop / Mobile) and view mode (Reading / Editing).

Amazing news, step by step you are creating a powerhouse plugin. It put my eyes on version `1.5.15` and looks very good. 

It is amazing to see how the community is happy with the plugin. I knew it was gonna be a success !!!!

**Reply by chrisgurney** [reply]
@FelipeRearden Thanks again! Really appreciate the support!

**Reply by FelipeRearden** [reply]
Version `1.5.21` is working perfect on Obsidian Mobile !!! I really like `top(fixed)` option and the `*` all notes option.
- `Top(fixed)` is working perfect on Reading Mode #20 (as you already know); 
- Style Settings are great and will help a lot of users;
- `*` is respecting specific folder and yaml settings;
- Properties cmds are AMAZING and I am gonna use these commands for  a new way to intereact with YAML from keyboard (wip)

I wish you a fantastic sunday ☀️

**Comment by chrisgurney** [comment]
Feedback on this idea from another user:

> I prefer source mode bc with mobile preview is problematic for me. It's hard to get the cursor at the correct position to edit certain parts or even see that there is code hidden. Downside is source mode is cluttered with things that are there for technical reasons and not the actual purpose of my note. This applies to my fleeting notes, my knowledge section is rather clean.
> 
> Scenarios for both preview and source where I can imagine wanting to have a different toolbar:
> 
> - for navigating my note: jump to predefined places/ headers/ dataview-fields/ table cells/ anchors in general. No more scrolling and trying to tap at the exact spot...
> - toggle some previews like code sections, images, transclusions, etc.
> - refract the note
> - general formatting (hate the formatting toolbar bc they are not contextual!)
> - in general calling functions or templates that are relevant when initially creating the note vs. later reading the note and work with it.
> 
> I hope I could articulate clearly what potential I see. These are not features that I desperately need, and yet they can come in handy and I am sure there are lots of other ideas out there.

Source: https://www.reddit.com/r/ObsidianMD/comments/1covqha/comment/l3h1m9v/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

**Comment by chrisgurney** [comment]
Adding to this idea, from another user:

> Would it be possible to make items visible/hidden based on more than just desktop/mobile?
> For example, if the file is not a markdown file but a canvas/multimedia, some items can be invisible.

**Comment by FelipeRearden** [comment]
Hello @chrisgurney !!!!

Based on the new feature from 1.24-beta…

> Setting for Command items: Hide if not available
> 
> Enable this setting to hide the command item if the command cannot be performed in the current context. Disabled by default.

… I have a question to you :)

Would you mind checking if the structure made to this feature couldn’t be used to add options to show NTB items only in the Reading Mode /Live Preivew? 

In my specific case, almost all of my NTB are items are commands from  templater scripts. This way, this items will alway be visible. 

Would be amazing to be able to have a filter for Reading Mode 🙏☀️

**Reply by chrisgurney** [reply]
@FelipeRearden 

> Would you mind checking if the structure made to this feature couldn’t be used to add options to show NTB items only in the Reading Mode /Live Preivew?

Unfortunately no. This new setting just uses an API another user made be aware of: https://docs.obsidian.md/Plugins/User+interface/Commands#Conditional+commands

Your request is still in my head, and the settings data structure should support it. I think the only challenge will be expanding the UI without making it more complex.

**Reply by FelipeRearden** [reply]
> Unfortunately no. This new setting just uses an API another user made be aware of: https://docs.obsidian.md/Plugins/User+interface/Commands#Conditional+commands

Thanks for trying @chrisgurney  and for the explanation!

> Your request is still in my head, and the settings data structure should support it. I think the only challenge will be expanding the UI without making it more complex.

Happy to know that you have eyes on this feature 🙏 

I think if we have `Hide if not available` with a new option to `Show only in Reading Mode` we close the output options that makes the NTB dynamic.

I wish you a fantastic day ☀️

**Comment by chrisgurney** [comment]
Adding another vote here for item visibility based on Editing vs. Reading vs. Source mode via @heyflorin in [#357](https://github.com/chrisgurney/obsidian-note-toolbar/issues/357):

> Additionally, if you have any separators added, they will remain on regardless. There should probably be a setting that simply turns the item on/off based on mode the user is in. In fact, the whole toolbar could have that option.

**Comment by Manrunas** [comment]
I was very happy after finding [obsidian-note-toolbar](https://github.com/chrisgurney/obsidian-note-toolbar) plugin. Started to use and almost immediately realized that it would be convenient to have separate toolbars for Editing and Reading. So I found this thread and it looks not only me who have same needs. 

I could use Reading View toolbar:
- folding/unfolding headings
- jumping to specific locations
- setting/toggling frontmatter properties
- resetting checklist in whole document at once

The other thing I found is that its only possible to map one toolbar per folder, now I assigned my main toolbar to all folders "\*" and I cannot map other toolbars to "\*". On mobile I would prefer to have 3 toolbars at the bottom: circle on the left, expanded bar in the center and circle on the right. : -) Moreover, if position coincide, for example 3 toolbars on the left, they could stack and appear as a column. Why not? :-)

**Reply by chrisgurney** [reply]
@Manrunas Thanks for the vote for this feature!

> Started to use and almost immediately realized that it would be convenient to have separate toolbars for Editing and Reading.

FYI the way I'm looking at implementing this is to make _items_ in toolbars separately configurable as to what's shown in Reading vs. Editing mode. (Like you can when you're choosing to show them on desktop, mobile, or both.)

As per your other request, supporting multiple toolbars per note is also something I'm considering for the future, but this is a more significant change to how the plugin currently works. 

Thanks also for your vote there as well.

**Comment by chrisgurney** [comment]
Hello @FelipeRearden @heyflorin and @Manrunas,

[In the latest beta](https://github.com/chrisgurney/obsidian-note-toolbar/blob/master/docs/releases/en/1.28.md), you can now control item visibility in Editing mode, Reading mode, or both (default). 🎉

Give it a try in the beta, [via BRAT](https://github.com/chrisgurney/obsidian-note-toolbar#installation), and let me know what you think!

> [!important]
> Back up your plugin's `data.json` file before using this beta, as it migrates your plugin data in order to support new item settings.

_Thank you for your patience in adding this feature, especially you @FelipeRearden!_

**Reply by FelipeRearden** [reply]
Hello @chrisgurney !!!!

Amazing news Chris! I am so happy that you was able to develop this feature 💙

Since my vault is not on `1.11` I was able to test `1.28.beta-2` in an Android phone on Obsidian Mobile `1.11.5`.  

**Working perfect on beta-2 on Obsidian Mobile**

The Only comment that I have about this feature is about a way to see which config each item has (Reading, Editing or Both) **without opening the Edit item window** 
- My suggestion is a colored items icon or border icon . I am assuming that is not feasible to add icons  in this screen (view image below)

📸 each square represents one of the 3 icons for each state (Readinf, Editing and both)
![IMG_0295](https://github.com/user-attachments/assets/89c35e0e-e283-4665-b1a5-1a1ff380d6de)

<hr>

P.S.: Thank you very much for treating the `separator item` as well. This was one of my concerning while testing, because my toolbars will use a lot of separators based on Reading/Editing states

**Reply by chrisgurney** [reply]
@FelipeRearden Thanks for the testing and feedback!

> ...a way to see which config each item has (Reading, Editing or Both) without opening the Edit item window

I'm open to ideas, but since visibility state is more than just about view mode: 

- The icon I might use to _summarize_ visibility of an item is an [eye icon](https://lucide.dev/icons/?focus=&search=eye) that represents whether the item's visible, not visible, or partially visible (I'll have to add a half-closed eye).
- On desktop, you could hover over such an icon to see all the details in a tooltip. 
- On mobile, there's no tootips, or room to show more information. But perhaps...
    - ...it would reflect the current visibility based on what platform you're on, and in what mode you're in, when viewing the items list. (I'd have to add a footnote or something similar to make that clear.)
    - Unfortunately you would have to tap on the item to see the applied visibility settings.

**Reply by FelipeRearden** [reply]
100% understood the issue we have here. Some insight…

The ideal world would be display all states in a single row (for desktop and mobile), like this:

![IMG_0296](https://github.com/user-attachments/assets/551d1610-a51c-4c96-b4e6-eaada00c906a)

Since this approach is not feasible, I think your idea is great!!!

I wish you a fantastic day ☀️

**Reply by chrisgurney** [reply]
Hey @FelipeRearden try [the latest beta](https://github.com/chrisgurney/obsidian-note-toolbar/blob/master/docs/releases/en/1.29.md).

I've added visibility indicators to each row, but I limit it to one. So, if the item can't be seen on the current platform, there's a platform-related indicator. If it's visible on both platforms, and it's then limited to a specific mode, then the mode indicator is shown. Indicators are only shown based on the mode of the currently-opened note.

**Reply by FelipeRearden** [reply]
Thank you for the new beta @chrisgurney !!!

Working perfect on version `1.29-beta-3`

> I've added visibility indicators to each row, but I limit it to one. So, if the item can't be seen on the current platform, there's a platform-related indicator. If it's visible on both platforms, and it's then limited to a specific mode, then the mode indicator is shown. Indicators are only shown based on the mode of the currently-opened note.

I tested on an Android phone and is working exactly the way that you mentioned!

I wish you a fantastic day ☀️

**Reply by chrisgurney** [reply]
Thanks @FelipeRearden 

What do you think of how it was implemented? Hopefully it's helpful and makes sense. If not, I'd love your feedback!

**Reply by FelipeRearden** [reply]
> What do you think of how it was implemented? Hopefully it's helpful and makes sense. If not, I'd love your feedback!

I think I found an issue… My first test was installing using BRAT (without Obsidian to the updating to the latest public release ).  Everything was okay.

But after updating Obsidian, I got this screen:

![IMG_0526](https://github.com/user-attachments/assets/1de53514-6a15-4384-966c-b423562538fc)

And this error:

![IMG_0527](https://github.com/user-attachments/assets/21032d14-246a-4ca8-9b33-c4d9424eee20)

**Reply by chrisgurney** [reply]
@FelipeRearden Thanks for the error output; very helpful! This should be fixed in the latest beta.

**Comment by FelipeRearden** [comment]
> @FelipeRearden Thanks for the error output; very helpful! This should be fixed in the latest beta.

🙏 thanks for the fix Chris! Working perfect on `beta-4`


> What do you think of how it was implemented? Hopefully it's helpful and makes sense. If not, I'd love your feedback!

> I've added visibility indicators to each row, but I limit it to one. So, if the item can't be seen on the current platform, there's a platform-related indicator. If it's visible on both platforms, and it's then limited to a specific mode, then the mode indicator is shown. Indicators are only shown based on the mode of the currently-opened note.

Let me see if I understood correctly:
First Rule
1️⃣ Platform icons: if I am on Mobile and the item is Desktop only it will show an icon `lucide-monitor`
2️⃣ Platform icons: if I am on Desktop and the item is Mobile only it will show an icon `lucide-table-smartphone`

When the first rule does not apply (visible in both platforms) whe have a Second Rule **based on hte mode of the opened note**

3️⃣ Mode icons:  opened note is `Reading Mode` and `Show only in Editing Mode` -> Show `lucide-pencil`
4️⃣ Mode icons:  opened note is `Reading Mode` and `Show only in Reading Mode` -> Show nothing

5️⃣ Mode icons:  opened note is `Editing Mode` and `Show only in Editing Mode` -> Show nothing
6️⃣ Mode icons:  opened note is `Editing Mode` and `Show only in Reading Mode` -> Show `lucide-book`

If my matrix is right I **cannot** see 3️⃣ 6️⃣ . Even when I restart Obsidian and check again.

I am almost sure that I am doing something wrong :)

**Reply by chrisgurney** [reply]
@FelipeRearden 

> 3️⃣ Mode icons: opened note is Reading Mode and Show only in Editing Mode -> Show lucide-pencil

> 6️⃣ Mode icons: opened note is Editing Mode and Show only in Reading Mode -> Show lucide-book

> If my matrix is right I cannot see 3️⃣ 6️⃣ . Even when I restart Obsidian and check again.

Do you have a note currently open, in the mode where the toolbar item should _not_ appear?

**Reply by chrisgurney** [reply]
@FelipeRearden I've also now updated the icons to reflect the same state as the visibility icons themselves. Hopefully that's a little more intuitive for people.

**Reply by FelipeRearden** [reply]
> Do you have a note currently open, in the mode where the toolbar item should not appear?

Yeah! I am doing what I think is the standard:
- **note is open where the toolbar item should not appear**
- swipe top to bottom to open command palette
- tap Open  Settings
- go to NTB settings 
- do the modes change

I still get `no icon` in 3️⃣ and 6️⃣. Looks like that I don’t have any note open before going to NTB settings.

My opinion: since it is working on your end there is a huge change that is my bad skill at using Obsidian with fingers on a phone. It is really uncomfortable to use obsidian with fingers instead mouse and keyboard.  Maybe another Android power user could give us some positive feedback.
- My fear is to say that there is something wrong where is not wrong due to my poor knowledge with this OS.

Another thing: it is possible that this new obsidian version has same error on Android that could be  jeopardize our tests.

**Reply by chrisgurney** [reply]
@FelipeRearden This is the code I'm using to get the markdown view before I can get its mode. Perhaps this is working different for you (returning `undefined`), and/or I'm wondering if multiple views/tabs open might have an effect... or perhaps there's a more reliable way to do this:

```js
/**
 * Gets the view mode of the latest markdown view.
 * @returns MarkdownViewModeType or undefined, if not available.
 */
getLatestViewMode(): MarkdownViewModeType | undefined {
  let latestMode: MarkdownViewModeType | undefined = undefined;
  const leaves = this.ntb.app.workspace.getLeavesOfType('markdown');
  const activeView = leaves.length > 0 ? leaves[0].view as MarkdownView : null;
  if (activeView && activeView instanceof MarkdownView) {
    latestMode = activeView.getMode();
  }
  return latestMode;
}
```

**Reply by FelipeRearden** [reply]
> @FelipeRearden This is the code I'm using to get the markdown view before I can get its mode. Perhaps this is working different for you (returning undefined), and/or I'm wondering if multiple views/tabs open might have an effect... or perhaps there's a more reliable way to do this:

**I think I found something !!!!!!**

> I'm wondering if multiple views/tabs open might have an effect...

You got me on this quote … I had 3 open notes/tabs all this time. I closed the 2 others and let only one. **IT IS WORKING NOW !!!!!**

![IMG_0530](https://github.com/user-attachments/assets/2a44ce08-e3fc-4f2e-8040-d72a5378850a)


Maybe you are right, we have some issue with more than one tab open. Let me know if want me to test something else!

**Reply by chrisgurney** [reply]
@FelipeRearden OK, thanks! I'll try to reproduce on my (iOS) phone, and confirm if using the first returned leaf (`leaves[0].view`) is appropriate here.

**Reply by chrisgurney** [reply]
@FelipeRearden OK I think I've got a fix in place! Thanks again @FelipeRearden !

**Reply by FelipeRearden** [reply]
Working perfect on `beta-6` Chris!!!!

Tested with 1 Tab, 2 Tabs and 3 tabs switching from Reading mode and Live Preview.

**Reply by chrisgurney** [reply]
Great, thanks again @FelipeRearden !
